---
title: INTRODUCTION TO POPULATION GENETICS IN R
author: Alexa Resler
output: html_document
---


```{r, include=F}
# Set working directory
setwd("/Users/alexaresler/Dropbox/Talks/R Genetics/Data")
#setwd("C:/Users/alexa/Dropbox/Talks/R Genetics/Data")
wd = getwd()
```


# Genetic association study example using R

## Load packages
```{r, echo=T, message=F, warning=F}
library(SNPassoc)
library(genetics)
library(GenABEL)
library(qqman)
library(ggplot2)
```


## Import data
```{r}
data = read.csv("Data PH SNP.csv", 
         header = T, 
         na.strings = "NA")

data.snp.info = read.csv("Data SNP INFO.csv", 
                         header = T, 
                         na.strings = "NA")
```

```{r, echo=T}
head(data[1:10])
```

```{r, echo=T}
head(data.snp.info)
```


## SNP QC

Genotype frequency 

```{r, echo=T}
summary(data$rs10)
geno.freq = as.vector(summary(data$rs10) / sum(!is.na(data$rs10)))
geno.freq
```


Minor allele frequency

```{r, echo=T}
geno.freq[1]+(geno.freq[2]/2)
```


Major allele frequency

```{r, echo=T}
geno.freq[3]+(geno.freq[2]/2)
```


Call rate

```{r, echo=T}
sum(!is.na(data$rs10)) / nrow(data)
```


Examine genotype and allele frequencies after converting SNP to a genotype object

```{r, echo=T}
snp.summ = summary(genotype(data$rs10))
```


Genotype frequency

```{r, echo=T}
snp.summ$genotype.freq
```


Allele frequency

```{r, echo=T}
snp.summ$allele.freq
```


Number of samples typed

```{r, echo=T}
snp.summ$n.typed
```


Total number of samples

```{r, echo=T}
snp.summ$n.total
```


Call rate 

```{r, echo=T}
snp.summ$n.typed / snp.summ$n.total
```


Hardy-Weinberg equilibrium (HWE)

- Null hypothesis: HWE holds

Exact test 

```{r, echo=T}
snp.hwe.e = HWE.exact(genotype(data$rs10))
snp.hwe.e$p.value
```

Chi-Square Test 

```{r, echo=T}
snp.hwe.c = HWE.chisq(genotype(data$rs10))
snp.hwe.c$p.value
```


Checking MAF, call rate, & HWE for a large number of SNPs

```{r, echo=T}
check.snp = function(SNP){
  snp.summ = summary(genotype(SNP))
  snp.info = c(snp.summ$allele.freq[,2][2], 
               snp.summ$n.typed / snp.summ$n.total,
               HWE.exact(genotype(SNP))$p.value)
  names(snp.info) = c("maf", "call.rate", "hwe.exact.p")
  return(snp.info)
}

check.snp(data$rs10)
```

```{r, echo=T}
snp.qc = data.frame(t(apply(data[,6:55], 2, 
                            function(x) check.snp(x))))
head(snp.qc)
```


SNPs with a call rate < 95%

```{r, echo=T}
snp.qc[which(snp.qc$call.rate<0.95),]
```


SNPs with a MAF < 5%

```{r, echo=T}
snp.qc[which(snp.qc$maf<0.05),]
```


SNPs out of HWE (HWE p < 0.001)

```{r, echo=T}
snp.qc[which(snp.qc$hwe.exact.p<0.001),]
```


Exclude SNPs that did not pass QC checks

```{r, echo=T}
snps.exl = c(rownames(snp.qc[which(snp.qc$call.rate<0.95),]),
             rownames(snp.qc[which(snp.qc$maf<0.05),]),
             rownames(snp.qc[which(snp.qc$hwe.exact.p<0.001),]))
snps.exl = snps.exl[!duplicated(snps.exl)]
snps.exl

data.qc = subset(data, select = -which(colnames(data) %in% snps.exl))
```


GenABEL package

```{r, echo=T, eval=F}
check.marker(data, snpsubset, idsubset, 
             callrate = 0.95, maf = 0.05, p.level = 0.001) 
```


## Specify genetic models of inheritance

```{r, echo=T}
table(data$rs10)
```

Dominant

```{r, echo=T}
table(dominant(genotype(data$rs10)))
```


Recessive

```{r, echo=T}
table(recessive(genotype(data$rs10)))
```


Codominant

```{r, echo=T}
table(codominant(genotype(data$rs10)))
```


Additive

```{r, echo=T}
table(additive(genotype(data$rs10)))
```


## Association between SNP & outcome

Logistic regression model for a single SNP

```{r, echo=T}
mod = glm(case.ctrl ~ dominant(genotype(rs10)), 
          family=binomial(link = "logit"), data=data)
```

Create vector of results

```{r, echo=T, message=F}
mod.info = c(exp(c(mod$coef[2], 
                   confint(mod)[2,])), 
             car::Anova(mod, type="III")$"Pr(>Chisq)")
names(mod.info) = c("OR", "L", "U", "p")
mod.info
```


Logistic regression models for multiple SNPs

```{r, echo=T, message=F}
mod.snp = function(SNP, DATA){
  mod = glm(case.ctrl ~ dominant(genotype(SNP)), 
            family=binomial(link = "logit"), data=DATA)
  mod.info = c(exp(c(mod$coef[2], confint(mod)[2,])), 
               car::Anova(mod, type="III")$"Pr(>Chisq)")
  names(mod.info) = c("OR", "L", "U", "p")
  return(mod.info)
}

mod.snp(data.qc$rs10, data.qc)
```


```{r, echo=T, message=F}
mod.results = data.frame(t(apply(data.qc[,6:55], 2, 
                                 function(x) mod.snp(x, data.qc))))
head(mod.results)
```


## GWAS association

Logistic regression models for GWAS data

GenABEL package

```{r, echo=T, eval=F}
qtscore(case.ctrl~gender+age, data, trait.type="binomial",
        snps=c(1:100000))
```


GWASTools package

```{r, echo=T, eval=F}
source("https://bioconductor.org/biocLite.R")
biocLite("GWASTools")

assocRegression(data, outcome="case.ctrl", model.type="logistic",
                covar=c("gender", "age"), snpStart=1, snpEnd=100000)
```


## Multiple comparisons correction

Bonferroni

```{r, echo=T}
bonf.p = p.adjust(mod.results[,4], method = "bonferroni")
```


False discovery rate (FDR) 

```{r, echo=T}
fdr.p = p.adjust(mod.results[,4], method = "fdr")
```


Permutation tests


SNPassoc package

```{r, echo=T, eval=F}
permTest(x, method="minimum", K)
```

MCPerm package

```{r, echo=T, eval=F}
OR.TradPerm(genotypeLine, affectionLine, fromCol, 
            naString, sep, repeatNum = 1000)
```


## Visualizing results

Forest plot

```{r, echo=F}
knitr::opts_chunk$set(fig.width=8.2, fig.height=5.3) 
```

```{r, echo=F}
ggplot(mod.results, aes(y=OR, x = reorder(rownames(mod.results), OR))) +
  geom_point() +
  geom_errorbar(aes(ymin=L, ymax=U), width=.2) +
  geom_hline(yintercept = 1, linetype=2, color="red") +
  coord_flip() +
  labs(title = "", x = "SNP", y = "OR") +
  theme_bw() + 
  theme(axis.text.y=element_text(size=6)) 
```


P-value vs. SNP position

```{r, echo=F}
knitr::opts_chunk$set(fig.width=8, fig.height=5) 
```

```{r, echo=F}
data.lz = merge(mod.results, data.snp.info, by.x="row.names", by.y="SNP", all.x=T)
data.lz$p.trans = -log10(data.lz$p) 

ggplot(data.lz, aes(y=p.trans, x = Position)) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype=3, color="grey") +
  geom_text(data=subset(data.lz, p < 0.05),
            aes(Position, p.trans, label=Row.names), 
            hjust=-0.15 , color="red") +
  geom_point(data=subset(data.lz, p < 0.05),
             aes(Position, p.trans), color="red") +
  labs(title = "", x = "Position", y = "-log10(p)") +
  theme_bw()
```


## Visualizing GWAS results

Manhattan plot

```{r, echo=T}
nrow(gwasResults)

head(gwasResults)
```

```{r, echo=F}
knitr::opts_chunk$set(fig.width=8, fig.height=4.8) 
```

```{r, echo=T}
manhattan(gwasResults)
```


